verifier-sdk/src/light_transaction.rs

```diff
@@ -190,19 +193,80 @@ impl<
 
     #[heap_neutral]
     #[inline(never)]
-    pub fn emit_indexer_transaction_event(&self) -> Result<()> {
+    pub fn emit_indexer_events(&self) -> Result<()> {
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("pre assemble TransactionIndexerEvent");
 
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("pre load MerkleTreeSet");
-        let merkle_tree_set = self.input.ctx.accounts.get_merkle_tree_set().load()?;
+        let merkle_tree_set_info = self.input.ctx.accounts.get_merkle_tree_set();
+        let merkle_tree_set = merkle_tree_set_info.load()?;
         #[cfg(all(target_os = "solana", feature = "mem-profiling"))]
         custom_heap::log_total_heap("post load MerkleTreeSet");
 
-        // Initialize the vector of leaves
-        let first_leaf_index =
-            state_merkle_tree_from_bytes(&merkle_tree_set.state_merkle_tree).next_index;
+        let state_merkle_tree = state_merkle_tree_from_bytes(&merkle_tree_set.state_merkle_tree);
+
+        // Emit the changelog event, twice: for the left and the right leaf.
+
+        let changelog_index_right = state_merkle_tree.current_changelog_index;
+        let changelog_index_left = changelog_index_right.saturating_sub(1);
+
+        let path_left = state_merkle_tree.changelog[changelog_index_left as usize].path;
+        let path_left = path_left
+            .iter()
+            .enumerate()
+            .map(|(level, node)| PathNode {
+                node: node.to_owned(),
+                index: (1 << (path_left.len() as u32 - level as u32))
```

Comment:
@ananas-block Not gonna lie, I'm quite annoyed by this `as u32` conversion being repeated here, but I'm not sure
what we should do. The thing is that the `index` field in `ChangeLogEventV1` in SPL is u32 and they do the same stuff. I would like to do it better, here are the alternatives to consider:

1. Use `u32` as type for all indexed (`changelog_index`, `rightmost_index`, `current_index` etc.) in our concurrent Merkle tree. Any objections against this? This way, we would need only safe conversions from u32 to usize (`usize::from()` works fine in such case and never panics!) whenever we do for loops in Rust. Good benefit of that solution would be saving a bit of space in our MT accounts (not much, but still).
2. Use `u64` as the type in the event. Any objections about this? I don't think that indexers or clients would care much?

I'm leaning towards 1.

Also, we could radicalize the solution 1 even by using `u16`, which should be fine. We don't plan to store more roots or changelogs than 65_535, do we? :sweat_smile: In such case, we could safely convert it to u32 or usize whenever needed (on demand, when looping, sending event etc.).

Reply:
We are constraint by the height - number of leaves of the Merkle tree for the indexes thus we definitely cannot take u16.
u32 also just allows for trees of height 32 and we will want to do 40 for address spaces.
So the most general approach would be to use u64. However, 4 extra bytes per changelog add up a lot both in the account and for the indexers. If we can we should minimize the memory footprint for that.

Do you see a way how we could switch between u32 and u64?

A naive way would be to keep two implementations one for state Merkle trees with a max height of 32 and one for bigger Merkle trees.

Reply:
One way would be to have versioned events, something like:

```rust
pub enum ChangelogEvent {
    V1(ChangelogEventV1),
    V2(ChangelogEventV2),
}
```

Where the 1st one uses u32, the second one - i64. At the beginning, we would have only the V1 variant, then we could add V2 whenever we need

Reply:
That makes sense, would it make sense to count the current event as v1 and our as v2,v3?